packages/core/android/src/main/kotlin/com/datadog/reactnative/DdSdkBridgeExt.kt

```diff
@@ -6,44 +6,187 @@
-        @Suppress("NotImplementedDeclaration")
         when (v) {
             null -> map.putNull(key)
             is Int -> map.putInt(key, v)
             is Long -> map.putDouble(key, v.toDouble())
             is Float -> map.putDouble(key, v.toDouble())
             is Double -> map.putDouble(key, v)
             is String -> map.putString(key, v)
-            is List<*> -> map.putArray(key, v.toWritableArray())
-            is Map<*, *> -> map.putMap(key, v.toWritableMap())
-            else -> TODO()
+            is Boolean -> map.putBoolean(key, v)
+            is List<*> -> map.putArray(
+                key,
+                v.toWritableArray(
+                    createWritableMap,
+                    createWritableArray
+                )
+            )
+            is Map<*, *> -> map.putMap(
+                key,
+                v.toWritableMap(
+                    createWritableMap,
+                    createWritableArray
+                )
+            )
+            else -> Log.e(
+                javaClass.simpleName,
+                "Unhandled type ${v.javaClass.simpleName} has been ignored"
+            )
         }
     }
+
     return map
 }
+
+/**
+ * Recursively converts the [ReadableMap] to a [Map] which only contains Kotlin stdlib objects,
+ * such as [List], [Map] and the raw types.
+ */
+internal fun ReadableMap.toMap(): Map<String, Any> {
+    val map: HashMap<String, Any> = this.toHashMap()
+    val iterator: Iterator<String> = map.keys.iterator()
+
+    while (iterator.hasNext()) {
+        val key = iterator.next()
+        try {
+            when (val type = getType(key)) {
+                ReadableType.Map -> {
+                    map[key] = getMap(key)?.toMap() ?: Log.e(
+                        javaClass.simpleName,
+                        "Cannot convert nested map for key: $key"
+                    )
+                }
+
+                ReadableType.Array -> {
+                    map[key] = getArray(key)?.toList() ?: Log.e(
+                        javaClass.simpleName,
+                        "Cannot convert nested array for key: $key"
+                    )
+                }
+
+                ReadableType.Null, ReadableType.Boolean, ReadableType.Number, ReadableType.String -> {}
+                else -> Log.e(
+                    javaClass.simpleName,
+                    "Skipping unhandled type [${type.name}] for key: $key"
+                )
+            }
+        } catch (err: IllegalArgumentException) {
+            Log.e(javaClass.simpleName, "Could not convert object for key: $key.", err)
+        }
+    }
+
+    return map
+}
+
+/**
+ * Recursively converts the [ReadableArray] to a [List] which only contains Kotlin stdlib objects,
+ * such as [List], [Map] and the raw types.
+ * or [List], instead of [ReadableMap] and [ReadableArray] respectively).
+ */
+internal fun ReadableArray.toList(): List<*> {
+    val out = mutableListOf<Any?>()
```

Comment:
better to use non-nullable type here, because `null` values there are useless and it only widens the type (you need to handle null values) later when caller iterates over it.

Also it is better to rename it, because `out` is a reserved keyword in Kotlin (even here syntax highlighting treats it differently).

```suggestion
    val list = mutableListOf<Any>()
```

Reply:
I am supporting null values because ReadableArrays support them, and I wanted to provide a method to do a 1:1 conversion. What do you think?

I agree I should rename the variable.

Reply:
well, down the line you are calling `mapNotNull` anyway and as I replied here https://github.com/DataDog/dd-sdk-reactnative/pull/644#discussion_r1549869773 probably the only reason `null` is supported in `ReadableArray` this weird use-case when it is used like a map and they need to keep structural ordering.

But I'm fine with keeping it as `<Any?>`, just wanted to simplify it a bit (and having `null` values in the list is not quite idiomatic in Java/Kotlin, it makes little sense).

Reply:
At the moment `mapNotNull` is used in one specific use-case (`asFirstPartyHosts()`), but the extension method is generic and it will surely be used many times in the future.

I agree that in 99% of real use cases, having null values makes little sense.

On the other hand, if for some reason I want to return 3 arguments from Kotlin to the RN side, and I want them to be nullable and keep their position in the array, I'd have a list like this:

```kotlin
val args = listOf(null, "arg2", "arg3")
```

And I would convert it to a `WritableArray`, expecting to find the elements in the same position, so I'd call `args.toWritableArray()` and get a perfect match.

On the other hand, if I am aware my list can contain null values, and I want to remove them, then I'd call `args.mapNotNull().toWritableArray()`.

I think this way we ultimately leave the choice to the developer.